rm/rm-server/src/main/java/org/ow2/proactive/resourcemanager/core/NodesRecoveryManager.java

```diff
@@ -75,8 +98,190 @@ public NodesLockRestorationManager apply(RMCore rmCore) {
         };
     }
 
-    public void restoreLocks(RMNode rmNode, Client caller) {
-        nodesLockRestorationManager.handle(rmNode, caller);
+    public void recoverNodeSourcesAndNodes() {
+        Collection<NodeSourceData> nodeSources = rmCore.getDbManager().getNodeSources();
+        logPersistedNodeSourceInfo(nodeSources);
+
+        for (NodeSourceData nodeSourceData : nodeSources) {
+            String nodeSourceName = nodeSourceData.getName();
+            if (NodeSource.DEFAULT_LOCAL_NODES_NODE_SOURCE_NAME.equals(nodeSourceName)) {
+                // will be recreated by SchedulerStarter
+                rmCore.getDbManager().removeNodeSource(nodeSourceName);
+            } else {
+                recoverDeployedNodeSourceIfNeeded(nodeSourceData, nodeSourceName);
+            }
+        }
+    }
+
+    public boolean recoverInfrastructureVariablesIfNeeded(String nodeSourceName, NodeSource nodeSourceToDeploy,
+            NodeSourceDescriptor descriptor) {
+        boolean existPersistedNodes = existPersistedNodes(nodeSourceName, nodeSourceToDeploy.nodesRecoverable());
+        boolean recoverNodes = existPersistedNodes;
+
+        if (existPersistedNodes) {
+            InfrastructureManager im = InfrastructureManagerFactory.recover(NodeSourceData.fromNodeSourceDescriptor(descriptor));
+            if (!im.getDeployingAndLostNodes().isEmpty()) {
+                // if there are deploying nodes, we will not recover
+                rmCore.getDbManager().removeAllNodesFromNodeSource(nodeSourceName);
+                recoverNodes = false;
+            } else {
+                nodeSourceToDeploy.setInfrastructureManager(im);
+            }
+        }
+        return recoverNodes;
+    }
+
+    private boolean existPersistedNodes(String nodeSourceName, boolean nodesRecoverable) {
```

Comment:
`private boolean existPersistedNodes(String nodeSourceName, boolean nodesRecoverable)`

You can change the method signature to `private boolean existPersistedNodes(String nodeSourceName)`

Because when `nodesRecoverable`is false this `(PAResourceManagerProperties.RM_NODES_RECOVERY.getValueAsBoolean() && nodesRecoverable)` evaluates to false. So the method returns false. 
If `nodesRecoverable` is true, then the method is executed and validates as true or false.